### DISCUSSION TITLE
I think _update_metadata needs a major rewrite

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#226] [Design & Development]
This was precipitated by a problem I'm encountering in writing the material for the upcoming sage/gage short course where we are going to roll out MsPASS to the seismology community.  The "bottom line" to use an overused cliche is the way we are handling normalized attributes is most definitely broken.   Any time I do a read like this one:
```
normalize_list=['site','source']
d_normalized=dbh.read_data(doc,collection='wf_Seismogram',normalize=normalize_list)
```
and save it later like this:
```
dbh.save_data(d_normalized,collection='wf_Seismogram',data_tag='schema_feature_demonstration')
```
and then look at the contents of the document this wrote in wf_Seismogram  like this:
```
doc=dbh.wf_Seismogram.find_one({'data_tag' : 'schema_feature_demonstration'})
print(json_util.dumps(doc,indent=5))
```
I get some of this kind of problem children in the output:
```
   "READONLYERROR_net": "TA",
   "READONLYERROR_site_elev": 0.062,
   "READONLYERROR_site_endtime": 1470873599.0,
   "READONLYERROR_site_lat": 33.257702,
   "READONLYERROR_site_lon": -92.803001,
   "READONLYERROR_site_starttime": 1299715200.0,
   "READONLYERROR_source_depth": 53.2,
   "READONLYERROR_source_lat": 38.2525,
   "READONLYERROR_source_lon": 141.73,
   "READONLYERROR_source_magnitude": 7.1,
   "READONLYERROR_source_time": 1302186764.1,
   "READONLYERROR_sta": "Z41A"
```
I iniitially suspected a problem with read_data not calling the clear_modified method marking the normalized attributes changed, but digging showed that I don't think that is the case (I'm not sure).   What is really clear to me when I dig into this is I think we have a mess in the Database class method called _update_metadata.   That has been patched and patched and is now a python version of spaghetti code.  The proof is that the function has multiple boolean flags (has_fatal_error, new_insertion, and possibly more) and numerous (I count 4) continue statements - the closest thing python has to a goto.   

Now I started to work on trying to debug the specific problem noted above, and I decided we should seriously consider cleaning up that function with a complete redesign to clean up the logic?  

One thing I think would help is if we had two separate methods to handle the case of an actual metadata (document) update versus the logic you get in saving a mspass data object.   I actually think the need for an internal _update_metadata method is, in fact, questionable.   I think it is a relic of the evolution of this code we should rethink.   I think we should consider moving all _update_metadata code to save_data.   I think we also need to clean up the logic currently handled by the "new_insertion" boolean.  We should either consider a different method with an independent name (e.g. save_raw_data)  for saving raw data (no schema at all)  or have a new argument that defines the "new_insertion" logic.   

I don't want to just complain about this. This fix is important and if you concur I will dive into the problem and build a revision on a separate branch for the group's consideration.   One reason for starting this discussion is I think @haruming may have some fixes that aren't checked in yet.   If so, please advise where I can see them.  If not, I can proceed from the copy of master I refreshed just this morning.  

Either way @wangyinz I need a green light to go ahead with this or tell me why I shouldn't?  

Sorry this came off sounding a bit grumpy.   It shows my frustration in trying to follow the logic of _update_metadata

**Comment by wangyinz** [comment]
Yes, I agree the save related methods need a clean up. It got more and more complex compared to the original implementation, and it is indeed difficult to read now. We should absolutely revise that.

In terms of the problem, it is actually irrelevant to the structure of the code, but actually more of a behavior issue. if you remember, in the original version, we don't have this behavior of adding `READONLYERROR` to read-only keys, instead, read-only keys will just be ignored in a save. We later added this feature following your advise and a lot of discussions over the design of dbclean and dbverify method. Because the save method is trying to preserve everything, it now saves the read-only keys as `READONLYERROR`. I guess what you are trying to say here is that the original behavior is preferable, and I think that make sense. Read-only is read-only, and there is no reason to save it to the database. Also, another solution is what we have discussed in #199 - you could use a schema that does not have those keys defined read-only. I think using a different schema is not really a good solution, we should just reverse the change to that behavior, and ignore read-only keys in the save and update methods.

**Reply by wangyinz** [reply]
Wait, you mentioned the `clear_modified` method, but isn't that irrelevant? The save method will always save all the keys regardless it is modified or not.

**Comment by haruming** [comment]
Yes I do think our reader and writer right now are way more complex than before and we do need to revise and reorganize them. Otherwise, it may be more difficult to debug when bad things happen. One thing I could come up with is that we could split the logic of the code into several independent functions as you suggested. The more important question is do we have to revise it right now or later?
For the problem, I agree with what Ian said and this is align with the documentation we discussed before. And I understand that those READONLYERROR attributes are somehow irrelevant to the workflow and we could reverse it, or maybe use a boolean variable to control the behavior. I am open to either ways.

**Comment by pavlis** [comment]
Ok, guys. I'm on this.   I'm far enough already I'm sure this will be a big improvement.  It is also clear in working through this that we had created some unnecessary inefficiencies.   The most significant were multiple queries done during a save.   I think I have a way to work around that, but will know more later.   Thanks for the feedback.

**Comment by pavlis** [comment]
Can one of you guys explain to me why the following incantation at the end of _update_metadata is present:
```
        else:
            # FIXME: we could have recorded the full stack here, but need to revise the logger object
            # to make it more powerful for Python logging.
            mspass_object.elog.log_verbose(
                sys._getframe().f_code.co_name, "Skipped updating the metadata of a dead object")
            self._save_elog(mspass_object)
```
google tells me the point is call to _getframe returns the function name that called _update_metadata.  Why?   I'm guessing this was a patch to allow _update_metadata to do only what it's name suggests.  Is that right?   

For the revision I'm working on _update_metadata is going away and will be replaced by a public method called update that has this signature:
```
def update(self,metadata,exclude_list=[],update_all=False):
```
This by default will call the modified method on Metadata, clear any keys found in exclude list, build an update record form the keys returned (fixing alias if applied), and apply it.  update_all would be bypass the exclude_list and modifed method logic and just blindly update the entire contents of metadata.

I think the new update is simpler and does more what we need.  Please advise if you have any suggested modifications as I'm about to code the new method

**Reply by pavlis** [reply]
Minor change when is started on this.  I decided to retain the old method name of update_metadata, BUT I am changing the function signature to be simpler as above.

**Comment by wangyinz** [comment]
The `sys._getframe()` call is used to get the function name of current function. It is just a better way than hard coding in the function name as "update_metadata" there. It has nothing to do with the `_update_metadata` method. This call will just return `_update_metadata` there.

There are two issues in the new method signature above. The first is that we should never define an empty list as the default parameter. It will produce some mysterious behavior between independent calls. You can find more details [here](https://stackoverflow.com/questions/366422/what-is-the-pythonic-way-to-avoid-default-parameters-that-are-empty-lists), but basically, you want to stick to the original - define the default as `None` and then check for it explicitly in the function body.

The second issue is that we probably shouldn't let `update_all` overrides `exclude_list`. It is possible that the user want to bypass the modified method, but in the meantime want to exclude certain keys. 

Actually, in terms of functionality, I don't see any difference between this new signature and the original (other than the renamed parameters). I think we still want to keep the parameters such as `collection` and `data_tag`. Also, are you suggesting removing the `mode` parameter? I wonder what's the default behavior for the inconsistent-from-schema keys.

**Reply by pavlis** [reply]
In fact, I'm largely reverting back to the original signature and just cleaning up the messy logic by separating the two methods completely.

**Comment by pavlis** [comment]
This maybe should have gone in email but can one of you guys explain to me what this little incantation is for that is in the current _update_metadata method of Database?
```
for k in copied_metadata:
                # HELP NEEDED:  I don't know what this is for
                if not str(copied_metadata[k]).strip():
                    copied_metadata.erase(k)
```
I'm guessing it is looking for empty string data such as what we often get for the loc code with miniseed data.  Is that right?

**Reply by pavlis** [reply]
Whether or not I'm right or wrong, that is a classic example of where a comment in the original code was needed.

**Comment by wangyinz** [comment]
You are right. `strip()` will remove spaces at the beginning or end of the string. The `if` statement is checking whether the string is empty. In Python, an empty string is boolean false.

**Comment by pavlis** [comment]
OK guys, I have a complete version of the rewrite of save_data and update_metadata.   The save_data method api has not changed much but I think it is now cleaner.  It does a version of what _update_metadata used to do in a private method handling errors more appropriately for a save. That is, it now follows the correct mspass approach of using elog to log any errors and never throwing an exception for anything but a usage error.  I am pretty sure the new implementation is much faster as it it requires fewer database transactions to complete because of the logic change I was able to make by putting the metadata save in the same function.  There is one change to save_data.   It now returns a list of three ids:  [wfid, history_id, elog_id] where some may be None.   I don't think that will cause trouble except, perhaps, if used in a map operator.  

The update_metadata method, on the other hand, has undergone a rather dramatic change.   Not as much as I proposed above, but there are still a lot of changes.  Probably the best way to tell you want they are is to show you the new docstring and ask for comments:
```
    def update_metadata(self, mspass_object, collection=None, mode='cautious',
                exclude_keys=None, alg_name='Database.update_metadata'):
        """
        Use this method if you want to save the output of a processing algorithm
        whose output is only posted to metadata.   That can be something as
        simple as a little python function that does some calculations on other
        metadata field, or as elaborate as a bound FORTRAN or C/C++ function
        that computes something, posts the results to Metadata, but doesn't
        actually alter the sample data.   A type example of the later is an amplitude
        calculation that posts the computed amplitude to some metadata key value.

        This method will ONLY attempt to update Metadata attributes stored in the
        data passed (mspass_object) that have been marked as having been
        changed since creation of the data object.  The default mode will
        check entries against the schema and attempt to fix any type
        mismatches (mode=='cautious' for this algorithm).  In cautious or
        pedantic mode this method can end up posting a lot of errors in
        elog for data object (mspass_object) being handled.  In
        promiscuous mode there are no safeties and the any values
        that are defined in Metadata as having been changed will be
        posted as an update to the parent wf document to the data object.

        :param mspass_object: the object you want to update.
        :type mspass_object: either :class:`mspasspy.ccore.seismic.TimeSeries` or :class:`mspasspy.ccore.seismic.Seismogram`
        :param exclude_keys: a list of metadata attributes you want to exclude from being updated.
        :type exclude_keys: a :class:`list` of :class:`str`
        :param collection: the collection name you want to use. If not specified, use the defined collection in the metadata schema.
        :param mode: This parameter defines how attributes defines how
          strongly to enforce schema constraints. As described above
          'promiscuous' justs updates all changed values with no schema
          tests.  'cautious', the default, enforces type constraints and
          tries to convert easily fixed type mismatches (e.g. int to floats
          of vice versa).  Both 'cautious' and 'pedantic' may leave one or
          more complaint message in the elog of mspass_object on how the
          method did or did not fix mismatches with the schema.  Both
          also will drop any key-value pairs where the value cannot be
          converted to the type defined in the schema.

        :type mode: :class:`str`
        :param alg_name: alg_name is the name the func we are gonna save while preserving the history.
        :type alg_name: :class:`str`
        :param alg_id: alg_id is a unique id to record the usage of func while preserving the history.
        :type alg_id: :class:`bson.objectid.ObjectId`
        :return:  integer count of the number of values altered.  -1 if the
          input was marked dead.
        """
```
Note the three big changes are:
1.  An update is defined in a completely different way.   readonly data are always excluded and only attributes Metadata says are in the modified list will be updated.  I had suggested the need for an update_all flag, but decided that was a bad idea for the way a method like this should be used.  If someone wants to 'shoot themselves in the foot' (to use another bad cliche) that way they can use the low level MongoDB api to do it and write their own function.  I don't think we should encourage it.  
2. Errors are handled differently.  Note I made cautious the default.  In that mode, it tries to fix type errors and does so silently if it succeeds.  Failures result only in a Complaint being posted to elog and the offending data is popped from the update dictionary.   (i.e. it refused to try to do an update).   Promiscuous does no testing of anything and just blindly updates all data that are (a) not readonly and (b) are defined as changed by the modified method of Metadata.   
3. This returns something different.  It used to return some error count.  Now it returns the number of key-value pairs actually updated.   I think that is a much more useful return.

**Reply by haruming** [reply]
Well done. Those look like reasonable changes. Could you help push your current working branch to the repo so that I could see if there is anything missing?

**Reply by pavlis** [reply]
I will shortly after I fix a few obvious problems revealed in running the tests.  I am going to need your help, @haruming , reworking the test_db.py script.   There is now a mismatch in what is tested where.  update_metadata test really don't work at all and some of the tests there should probably be done as extensions to tests of save_data.   I'm planning to fix what I can and comment out tests that I need your help fixing.   With a little luck that should be done this morning.

**Reply by haruming** [reply]
Cool. I can imagine that there would be many problems in running the tests and I will help to fix them.
